pkg/controllers/dnspolicy/dnspolicy_controller.go

```diff
@@ -61,26 +75,130 @@ func (r *DNSPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	dnsPolicy := previous.DeepCopy()
 	log.V(3).Info("DNSPolicyReconciler Reconcile", "dnsPolicy", dnsPolicy)
 
-	if err := r.ReconcileHealthChecks(ctx, log, dnsPolicy); err != nil {
+	markedForDeletion := dnsPolicy.GetDeletionTimestamp() != nil
+
+	targetNetworkObject, err := r.FetchValidTargetRef(ctx, dnsPolicy.GetTargetRef(), dnsPolicy.Namespace)
+	if err != nil {
+		if !markedForDeletion {
+			if apierrors.IsNotFound(err) {
+				log.V(3).Info("Network object not found. Cleaning up")
+				err := r.deleteResources(ctx, dnsPolicy, nil)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+			return ctrl.Result{}, err
+		}
+		targetNetworkObject = nil // we need the object set to nil when there's an error, otherwise deleting the resources (when marked for deletion) will panic
+	}
+
+	if markedForDeletion {
+		if controllerutil.ContainsFinalizer(dnsPolicy, DNSPolicyFinalizer) {
+			if err := r.deleteResources(ctx, dnsPolicy, targetNetworkObject); err != nil {
+				return ctrl.Result{}, err
+			}
+			if err := r.RemoveFinalizer(ctx, dnsPolicy, DNSPolicyFinalizer); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// add finalizer to the dnsPolicy
+	if !controllerutil.ContainsFinalizer(dnsPolicy, DNSPolicyFinalizer) {
+		if err := r.AddFinalizer(ctx, dnsPolicy, DNSPolicyFinalizer); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	err = r.reconcileResources(ctx, dnsPolicy, targetNetworkObject)
+	if err != nil {
```

Comment:
Added ready condition for policy and not update status regardless of error during reconciliation. Further enhancements required here to deal with health check status, and making sure we always process the status (Missing target for instance still skips the status update) but this will cover most issues that can happen during reconciliation.